h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -0,0 +1,122 @@
+package water;
+
+import hex.Model;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.util.Log;
+
+import java.util.*;
+
+public class DKVManager {
+
+  /**
+   * Clears keys in all H2O nodes, except for the ones marked as retained.
+   * Only Model and Frame keys are retained. If a key of any other type is provided, it will be removed as well.
+   *
+   * @param retainedKeys Keys of {@link Frame}s and {@link Model}s to be retained.
+   */
+  public static void retain(final Key[] retainedKeys){
+    final Set<Key> retainedSet = new HashSet<>(retainedKeys.length);
+    retainedSet.addAll(Arrays.asList(retainedKeys));
+    extractNestedKeys(retainedSet);
+
+    new ClearDKVTask(retainedSet.toArray(new Key[]{}))
+            .doAllNodes();
+  }
+
+  /**
+   * Iterates through the keys provided by the user, dropping any keys that are not a Model key or a Frame key.
+   * Afterwards, extracts
+   *
+   * @param retainedKeys A {@link Set} of retained keys to insert the extracted {@link Frame} and {@link Model} keys to.
+   *                     Should contain user-specified keys to retain in order to extract anything.
+   */
+  private static void extractNestedKeys(final Set<Key> retainedKeys) {
+
+    final Iterator<Key> keysIterator = retainedKeys.iterator(); // Traverse keys provided by the user only.
+    while (keysIterator.hasNext()) {
+      final Key key = keysIterator.next();
+
+      final Value value = DKV.get(key);
+
+      if (value == null || value.isNull()) {
+        continue; // The key might not have value cached locally. Do not delete.
+      } else if (!value.isFrame() && !value.isModel()) {
+        Log.info(String.format("Given key %s is of type %d. Not retaining.", key.toString(), value.type()));
```

Comment:
perhaps a warning would be better?

Reply:
or should we just fail?

Reply:
We might as well fail and the user will know what he or she did wrong.

Reply:
Resolved in https://github.com/h2oai/h2o-3/pull/3577/commits/3786ef7de44574103c72247aba21c48e07126e67

Reply:
Also added a test https://github.com/h2oai/h2o-3/pull/3577/commits/ee15d8a864bc3ec558cbb10329e6d38cb5aa781f